File: js/recursion/inout/permutation.js
export const printPermutationWithSpaces = (inp, out) => {
  if (inp.length == 0) {
    console.log(out);
    return;
  }
  let op1 = out;
  let op2 = out;
  op1 = op1 + inp.charAt(0);
  op2 = op2 + " " + inp.charAt(0);
  inp = inp.substring(1);
  printPermutationWithSpaces(inp, op1);
  printPermutationWithSpaces(inp, op2);
};

export const printPermutationWithChange = (inp, out) => {
  if (inp.length == 0) {
    console.log(out);
    return;
  }
  let op1 = out;
  let op2 = out;
  op1 = op1 + inp.charAt(0);
  op2 = op2 + inp.charAt(0).toUpperCase();
  inp = inp.substring(1);
  printPermutationWithSpaces(inp, op1);
  printPermutationWithSpaces(inp, op2);
};

File: java/recursion/src/com/recursion/inout/Permutation.java
package com.recursion.inout;

public class Permutation {

    public void printPermutationWithSpaces(String in, String out) {
        if (in.length() == 0) {
            System.out.println(out);
            return;
        }
        String op1 = out;
        String op2 = out;
        op1 = op1.concat(String.valueOf(in.charAt(0)));
        op2 = op2.concat(" " + in.charAt(0));
        in = in.substring(1);
        printPermutationWithSpaces(in, op1);
        printPermutationWithSpaces(in, op2);
    }

    public void printPermutationWithChange(String in, String out) {
        if (in.length() == 0) {
            System.out.println(out);
            return;
        }
        String op1 = out;
        String op2 = out;
        op1 = op1.concat(String.valueOf(in.charAt(0)));
        op2 = op2.concat(String.valueOf(in.charAt(0)).toUpperCase());
        in = in.substring(1);
        printPermutationWithChange(in, op1);
        printPermutationWithSpaces(in, op2);
    }

}
